### PR TITLE
`fedimintd dev list-versions` subcommand

### PR DESCRIPTION
**Pull request by emilioziniades**
The server-side work for #4584.

Creates an optional subcommand `dev` for fedimintd, similar to fedimint-cli. 

~~Adds a single subcommand to `dev`: `list-versions`.This dumps all API and database versions and exits early.~~ 
Adds two subcommands to `dev`: `list-api-versions` and `list-db-versions`. These dump API and database versions respectively and exit early.

e.g.
```
FM_BITCOIN_RPC_KIND="bitcoind" FM_BITCOIN_RPC_URL="http://localhost:5002" fedimintd dev list-versions
```

<details>
<summary>And here's a sample of the output</summary>

```
{
  "core_api_versions": [
    {
      "major": 0,
      "minor": 2
    }
  ],
  "modules": [
    {
      "kind": "ln",
      "api_versions": [
        {
          "major": 0,
          "minor": 1
        }
      ],
      "database_version": 0
    },
    {
      "kind": "meta",
      "api_versions": [
        {
          "major": 0,
          "minor": 0
        }
      ],
      "database_version": 0
    },
    {
      "kind": "mint",
      "api_versions": [
        {
          "major": 0,
          "minor": 0
        }
      ],
      "database_version": 0
    },
    {
      "kind": "wallet",
      "api_versions": [
        {
          "major": 0,
          "minor": 0
        }
      ],
      "database_version": 0
    }
  ]
}
```

</details>

Unfortunately, the FM_BITCOIN_RPC_KIND and FM_BITCOIN_RPC_URL environment variables still need to be provided because otherwise fedimintd throws an error in `Fedimintd::new`.

It seems weird that they aren't part of the explicit CLI options but rather grabbed from the environment later:
https://github.com/fedimint/fedimint/blob/11c24a949840b317d52f85e282c1d43cabb2bafe/fedimintd/src/fedimintd.rs#L182

I did consider refactoring this into the `ServerOpts` struct (clap already supports reading from environment variables), but this doesn't resolve the issue that they are still required fields.

Basically I'd love to be able to do `fedimint dev list-versions` without any additional args (cli or env-vars), but wasn't sure how to get there. And I thought I'd rather get some feedback on what I've done before proceeding.

### CODE REVIEW

**Review by elsirion — DISMISSED** [review]
tACK, idk if we really want to make it a `fedimintd` subcommand, but I remember @dpc discussing the direction to be taken in the issue, I'll leave it to you.

```
$ FM_BITCOIN_RPC_KIND="bitcoind" FM_BITCOIN_RPC_URL="http://127.0.0.1" cargo run --bin fedimintd -- dev list-versions
    Finished dev [unoptimized + debuginfo] target(s) in 0.19s
     Running `target-nix/debug/fedimintd dev list-versions`
{
  "core_api_versions": [
    {
      "major": 0,
      "minor": 2
    }
  ],
  "modules": [
    {
      "kind": "ln",
      "api_versions": [
        {
          "major": 0,
          "minor": 1
        }
      ],
      "database_version": 0
    },
    …
  ]
}
```

An alternative I could imagine: add the DB version (even though it's a `fedimintd` internal) to the `version` API endpoint.

```
$ fedimint-cli -- --data-dir cfg/… dev api version
{
  "value": {
    "core": {
      "core_consensus": {
        "major": 2,
        "minor": 0
      },
      "api": [
        {
          "major": 0,
          "minor": 1
        }
      ]
    },
    "modules": {
      "0": {
        "core_consensus": {
          "major": 2,
          "minor": 0
        },
        "module_consensus": {
          "major": 2,
          "minor": 0
        },
        "api": [
          {
            "major": 0,
            "minor": 0
          }
        ]
      },
      …
    }  
  }
}
```

**Comment by dpc** [comment]
Try to keep the format match the existing:

```
fedimint devimint> fedimint-cli discover-version
{
  "versions": {
    "core": {
      "major": 0,
      "minor": 2
    },
    "modules": {
      "0": {
        "major": 0,
        "minor": 1
      },
      "1": {
        "major": 0,
        "minor": 0
      },
      "2": {
        "major": 0,
        "minor": 0
      },
      "3": {
        "major": 0,
        "minor": 0
      }
    }
  }
}
```

in particular: modules should be a map (keyed on the `kind` like you do already, not an id, since we talk about static-support, not the config).

I'm guessing two sub-commands, one for db, the other for api will be the easiest to use for the consuming scripts.

> Unfortunately, the FM_BITCOIN_RPC_KIND and FM_BITCOIN_RPC_URL environment variables still need to be provided 

The pattern we use in `fedimint-cli` for same problem is making such fields `Option`, private, and add a `pub $name(&self) -> Result<$type>` getter that returns a descriptive error (`"$name needs to be provided"`). This way we only error out if the field is really needed.

**Review by dpc — APPROVED** [review]
LGTM, but please get one more opinion before landing.

**Comment by emilioziniades** [comment]
> > Unfortunately, the FM_BITCOIN_RPC_KIND and FM_BITCOIN_RPC_URL environment variables still need to be provided
> 
> The pattern we use in `fedimint-cli` for same problem is making such fields `Option`, private, and add a `pub $name(&self) -> Result<$type>` getter that returns a descriptive error (`"$name needs to be provided"`). This way we only error out if the field is really needed.

I considered this, but the problem is that the fields are required in `Fedimintd::new`, so they are needed early on in the initialization process.

https://github.com/fedimint/fedimint/blob/96b93f011815a22d1c022e4751167c2e87302834/fedimintd/src/fedimintd.rs#L183

Would sensible default values make sense here? e.g. FM_BITCOIN_RPC_KIND defaults to "bitcoind" and FM_BITCOIN_RPC_URL defaults to "http://localhost:8332"

**Comment by emilioziniades** [comment]
This is how the output looks for each subcommand:
```
$ FM_BITCOIN_RPC_KIND="bitcoind" FM_BITCOIN_RPC_URL="http://localhost:8332" cargo run --quiet --bin fedimintd -- dev list-api-versions
{
  "core": [
    {
      "major": 0,
      "minor": 2
    }
  ],
  "modules": {
    "ln": [
      {
        "major": 0,
        "minor": 1
      }
    ],
    "meta": [
      {
        "major": 0,
        "minor": 0
      }
    ],
    "mint": [
      {
        "major": 0,
        "minor": 0
      }
    ],
    "wallet": [
      {
        "major": 0,
        "minor": 0
      }
    ]
  }
}

$ FM_BITCOIN_RPC_KIND="bitcoind" FM_BITCOIN_RPC_URL="http://localhost:8332" cargo run --quiet --bin fedimintd -- dev list-db-versions
{
  "modules": {
    "ln": 0,
    "meta": 0,
    "mint": 0,
    "wallet": 0
  }
}
```

**Review by elsirion — APPROVED** [review]
